### PR TITLE
fix: revert&fix typing of di containers

### PR DESCRIPTION
**Pull request by bojiang**
## Checklist:

- [ ] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: scripts/tools/type_checker.sh [hunk 1]
```diff
@@ -1,3 +1,6 @@
 #!/usr/bin/env bash
 
-pyright --stats
```
Comment (parano) [inline scripts/tools/type_checker.sh:3]: the CI is using `scripts/ci/style/pyright_check.sh`, running in the checker docker image, not this file

Comment (bojiang) [reply scripts/tools/type_checker.sh:3]: yeah modifying this for local test

File: requirements/dev-requirements.txt [hunk 2]
```diff
@@ -11,3 +11,4 @@ types-setuptools
 types-PyYAML
 types-filelock
 attrs>=21.1
+pyright
```
Comment (aarnphm) [inline requirements/dev-requirements.txt:14]: rejected this

**Review by aarnphm — CHANGES_REQUESTED** [review]
remove uncorrect requirements

File: requirements/dev-requirements.txt (hunk 2 above)
Comment (aarnphm) [reply requirements/dev-requirements.txt:14]:
```suggestion
```

File: bentoml/_internal/configuration/containers.py [hunk 3]
```diff
@@ -174,7 +175,7 @@ def as_dict(self) -> t.Dict[str, t.Any]:
         return self.config
 
 
-@container
+@dataclass
```
Comment (aarnphm) [inline bentoml/_internal/configuration/containers.py:178]: I thought we stopped using dataclass?

**Comment by aarnphm** [comment]
@bojiang fixed to uses attrs instead of dataclass

File: bentoml/_internal/configuration/containers.py [hunk 4]
```diff
@@ -170,11 +171,11 @@ def override(self, keys: t.List[str], value: t.Any):
...
             ) from e
 
-    def as_dict(self) -> t.Dict[str, t.Any]:
-        return self.config
+    def as_dict(self) -> providers.ConfigDictType:
+        return t.cast(providers.ConfigDictType, self.config)
 
 
-@container
+@attr.s
```
Comment (sauyon) [inline bentoml/_internal/configuration/containers.py:178]: Use `attr.define` or pyright will complain.

File: bentoml/_internal/configuration/containers.py [hunk 5]
```diff
@@ -223,24 +224,21 @@ def tracer(
 
             return NoopTracer()
 
-    logging_file_directory = t.cast(
-        Provider[t.Callable[[], str]],
+    logging_file_directory = providers.Factory[str](
+        lambda default, customized: customized or default,
```
Comment (sauyon) [inline bentoml/_internal/configuration/containers.py:228]: we should decide whether we're ok with using `or` like this.

File: bentoml/_internal/configuration/containers.py [hunk 6]
```diff
@@ -223,24 +224,21 @@ def tracer(
...
         ),
+        config.logging.file.directory,
     )
 
 
 BentoMLContainer = BentoMLContainerClass()
 
 
-@container
+@attr.s
```
Comment (sauyon) [inline bentoml/_internal/configuration/containers.py:241]: Same as above

File: bentoml/_internal/configuration/containers.py [hunk 7]
```diff
@@ -315,8 +300,8 @@ def metrics_client(
         )
 
     # Mapping from runner name to RunnerApp file descriptor
-    remote_runner_mapping = t.cast(Provider[t.Dict[str, int]], providers.Static(dict()))
-    plasma_db: "PlasmaClient" = providers.Static(None)  # type: ignore[reportUnknownVariableType]
+    remote_runner_mapping = providers.Static[t.Dict[str, int]](dict())
+    plasma_db = providers.Placeholder["PlasmaClient"]()  # type: ignore
```
Comment (sauyon) [inline bentoml/_internal/configuration/containers.py:304]: Any reason this isn't just `t.Optional` and `None`?

File: scripts/ci/run_tests.sh [hunk 8]
```diff
@@ -41,9 +37,9 @@ getval(){
 run_python(){
   if [ "$USE_POETRY" -eq 1 ]; then
     need_cmd poetry;
-    poetry run python -m "$@";
+    poetry run python -m "$@"
```
Comment (sauyon) [inline scripts/ci/run_tests.sh:40]: We should remove this.

File: bentoml/_internal/configuration/containers.py (hunk 4 above)
Comment (aarnphm) [reply bentoml/_internal/configuration/containers.py:178]: attr.define broke the whole thing

File: bentoml/_internal/configuration/containers.py (hunk 6 above)
Comment (aarnphm) [reply bentoml/_internal/configuration/containers.py:241]: also doesn't seem like pyright is complaining

File: bentoml/_internal/configuration/containers.py (hunk 3 above)
Comment (parano) [reply bentoml/_internal/configuration/containers.py]: @aarnphm it is ok to use dataclass for this use case if it works ok - it does not add extra dependency, right?

File: bentoml/_internal/configuration/containers.py (hunk 5 above)
Comment (parano) [reply bentoml/_internal/configuration/containers.py:228]: I prefer specifically checking `is None`

File: bentoml/_internal/configuration/containers.py (hunk 7 above)
Comment (parano) [reply bentoml/_internal/configuration/containers.py:304]: I think for some of the configs, `None` actually means something, whereas `Placeholder` means user's config yaml did not touch this field

Comment (parano) [reply bentoml/_internal/configuration/containers.py:304]: but in this case, I think `t.Optional` and `None` is better. @bojiang any reason we are not doing that?

File: bentoml/_internal/configuration/containers.py (hunk 4 above)
Comment (aarnphm) [reply bentoml/_internal/configuration/containers.py]: nvm

File: bentoml/_internal/configuration/containers.py (hunk 3 above)
Comment (aarnphm) [reply bentoml/_internal/configuration/containers.py]: yep, just for consistency sake

File: bentoml/_internal/configuration/containers.py (hunk 4 above)
Comment (aarnphm) [reply bentoml/_internal/configuration/containers.py:178]: @bojiang so you suggest not to do this?

File: typings/schema.pyi [hunk 9]
```diff
@@ -2,18 +2,39 @@
 This type stub file was generated by pyright.
 """
 from typing import Any, Callable, Dict, Generic, List
```
Comment (aarnphm) [inline typings/schema.pyi:4]: ignore this file, I was meant to open a separate PR to address this

**Comment by aarnphm** [comment]
@bojiang good to merge?

**Review by aarnphm — APPROVED** [review]

File: bentoml/_internal/configuration/containers.py (hunk 4 above)
Comment (bojiang) [reply bentoml/_internal/configuration/containers.py:178]: just dataclass here works for both pyright and runtime. actually it should be `simple_di.container` (which is just an alias of `dataclass` in that lib but sadly pyright didn't understand it)

File: bentoml/_internal/configuration/containers.py (hunk 7 above)
Comment (bojiang) [reply bentoml/_internal/configuration/containers.py:304]: yeah Optional makes sense here. will modify it.

File: bentoml/_internal/configuration/containers.py (hunk 5 above)
Comment (bojiang) [reply bentoml/_internal/configuration/containers.py]: Actually I don't know why the default value get so twisted. But it's not about the current PR.

**Pull request merged**